### PR TITLE
[JSC] Add missing includes from implementation files that need them in order to compile on their own

### PR DESCRIPTION
**Pull request by akyrtzi**
#### 01ab6c22fca22b62dd438e6db59977647e720989
<pre>
[JSC] Add missing includes from implementation files that need them in order to compile on their own
<a href="https://bugs.webkit.org/show_bug.cgi?id=242757">https://bugs.webkit.org/show_bug.cgi?id=242757</a>

Reviewed by Justin Michaud.

Due to unified sources bundling there are implicit dependencies between implementation files and if
the files get compiled on their own it leads to compiler or linker errors, due to missing includes.

Add those missing includes so they can compile on their own, which is beneficial even for the unified build
because adding/removing files changes the unified source partitions and can expose these errors later on.

* Source/JavaScriptCore/API/JSAPIGlobalObject.mm:
* Source/JavaScriptCore/API/JSContext.mm:
* Source/JavaScriptCore/API/JSObjectRef.cpp:
* Source/JavaScriptCore/API/JSScript.mm:
* Source/JavaScriptCore/API/JSScriptSourceProvider.mm:
* Source/JavaScriptCore/inspector/remote/cocoa/RemoteInspectorXPCConnection.mm:
* Source/JavaScriptCore/jit/RegisterSet.cpp:
* Source/JavaScriptCore/wasm/js/WasmToJS.cpp:

Canonical link: <a href="https://commits.webkit.org/254924@main">https://commits.webkit.org/254924@main</a>
</pre>

https://github.com/False/commit/725e3ae8e67b69267b8d95dfb4adfb692e86a6ee

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/82892 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/69/builds/26872 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/14329 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/90964 "Built successfully") | [✅ 🛠 🧪 win](https://ews-build.webkit.org/#/builders/10/builds/144827 "Built successfully and passed tests") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/64/builds/25190 "Built successfully") | [✅ 🛠 mac-debug](https://ews-build.webkit.org/#/builders/71/builds/22772 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/75150 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/87545 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/88492 "Passed tests") | [❌ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/68/builds/20557 "Found 1 new test failure: fast/text/international/system-language/han-text-style.html (failure)") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/70520 "Built successfully") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/20912 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/75736 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/76449 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/64070 "Passed tests") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/1/builds/74074 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/66/builds/23378 "Built successfully") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/10390 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/38/builds/64394 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/67/builds/23316 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/11273 "Passed tests") | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/46/builds/24991 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/3439 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/63/builds/24827 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/60/builds/34404 "Passed tests") | [✅ 🛠 jsc-mips](https://ews-build.webkit.org/#/builders/37/builds/65093 "Built successfully") | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/65/builds/24793 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/30655 "Passed tests") | [✅ 🧪 jsc-mips-tests](https://ews-build.webkit.org/#/builders/45/builds/15762 "Passed tests") |

### CODE REVIEW

File: Source/JavaScriptCore/API/JSScript.mm [hunk 1]
```diff
@@ -24,12 +24,14 @@
  */
 
 #import "config.h"
+#import "JSBaseInternal.h"
```
Comment (aproskuryakov) [inline Source/JavaScriptCore/API/JSScript.mm]: This is a code style violation (files start with config.h and their own include), and thus suspicious. Also, we don't include JSBaseInternal.h essentially anywhere else, and JSBase.h includes are in headers.

But I'm not familiar with this code, so someone who works on JSC needs to weigh in.

File: Source/JavaScriptCore/API/JSScriptSourceProvider.mm [hunk 2]
```diff
@@ -24,6 +24,7 @@
  */
 
 #import "config.h"
+#import "JSBaseInternal.h"
```
Comment (justinmichaud) [inline Source/JavaScriptCore/API/JSScriptSourceProvider.mm]: ditto

File: Source/JavaScriptCore/API/JSScript.mm (hunk 1 above)
Comment (akyrtzi) [reply Source/JavaScriptCore/API/JSScript.mm]: This is needed to bring in the `JSC_OBJC_API_ENABLED` define, let me know if there's a better way.

File: Source/JavaScriptCore/API/JSScriptSourceProvider.mm (hunk 2 above)
Comment (akyrtzi) [reply Source/JavaScriptCore/API/JSScriptSourceProvider.mm]: This is needed to bring in the `JSC_OBJC_API_ENABLED` define, let me know if there's a better way.

File: Source/JavaScriptCore/API/JSScript.mm (hunk 1 above)
Comment (akyrtzi) [reply Source/JavaScriptCore/API/JSScript.mm]: (I changed it to `#import "JSBase.h"` which is the one defining `JSC_OBJC_API_ENABLED`)

`JSScriptInternal.h` has this as its first line:
```
#if JSC_OBJC_API_ENABLED
```
so the macro `JSC_OBJC_API_ENABLED` needs to be defined before including it.

File: Source/JavaScriptCore/API/JSScriptSourceProvider.mm (hunk 2 above)
Comment (akyrtzi) [reply Source/JavaScriptCore/API/JSScriptSourceProvider.mm]: (I changed it to `#import "JSBase.h"` which is the one defining `JSC_OBJC_API_ENABLED`)

`JSScriptSourceProvider.h` has this as its first line:
```
#if JSC_OBJC_API_ENABLED
```
so the macro `JSC_OBJC_API_ENABLED` needs to be defined before including it.

**Comment by akyrtzi** [comment]
Any comments or suggestions about this patch? 🙏

**Review by justinmichaud — APPROVED** [review]
r=me

**Pull request merged**